File: java/Samples/Biometrics/VB/AbisSampleVB/SubjectEditor/CaptureVoicePage.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Threading
Imports System.Windows.Forms
Imports Neurotec.Biometrics
Imports Neurotec.Biometrics.Client
Imports System.ComponentModel

Partial Public Class CaptureVoicePage
	Inherits Neurotec.Samples.PageBase
#Region "Public constructor"

	Public Sub New()
		InitializeComponent()

		AddHandler rbFromFile.CheckedChanged, AddressOf RadioButtonCheckedChanged
		AddHandler rbMicrophone.CheckedChanged, AddressOf RadioButtonCheckedChanged
	End Sub

#End Region

#Region "Private fields"

	Private _biometricClient As NBiometricClient
	Private _subject As NSubject
	Private _newSubject As NSubject
	Private _voice As NVoice
	Private _isIdle As New ManualResetEvent(True)
	Private _phrases As New List(Of Phrase)()

#End Region

#Region "Public methods"

	Public Overrides Sub OnNavigatedTo(ByVal ParamArray args() As Object)
		If args Is Nothing OrElse args.Length <> 2 Then
			Throw New ArgumentException("args")
		End If

		_subject = CType(args(0), NSubject)
		_biometricClient = CType(args(1), NBiometricClient)
		AddHandler _biometricClient.PropertyChanged, AddressOf OnBiometricClientPropertyChanged

		_newSubject = New NSubject()
		_voice = New NVoice()
		_newSubject.Voices.Add(_voice)
		voiceView.Voice = _voice
		_phrases = New List(Of Phrase)(SettingsManager.Phrases)
		ListAllPhrases()
		lblHint.Visible = False

		OnVoiceCaptureDeviceChanged()
		EnableControls()

		MyBase.OnNavigatedTo(args)
	End Sub

	Public Overrides Sub OnNavigatingFrom()
		If IsBusy() Then
			LongActionDialog.ShowDialog(Me, "Finishing current action ...", AddressOf CancelAndWait)
		End If
		RemoveHandler _biometricClient.PropertyChanged, AddressOf OnBiometricClientPropertyChanged
		If _voice.Status = NBiometricStatus.Ok Then
			Dim voices = _newSubject.Voices.ToArray()
			_newSubject.Clear()
			For Each item In voices
				_subject.Voices.Add(item)
			Next item
		End If
		voiceView.Voice = Nothing
		_newSubject = Nothing
		_voice = Nothing
		_subject = Nothing
		_biometricClient = Nothing

		MyBase.OnNavigatingFrom()
	End Sub

#End Region

#Region "Private methods"

	Private Sub SetHint(ByVal backColor As Color, ByVal format As String, ByVal ParamArray args() As Object)
		lblHint.BackColor = backColor
		lblHint.Text = String.Format(format, args)
		lblHint.Visible = True
	End Sub

	Private Sub ListAllPhrases()
		Try
			cbPhrase.BeginUpdate()
			Dim selected As Object = cbPhrase.SelectedItem
			cbPhrase.Items.Clear()
			For Each item In _phrases
				cbPhrase.Items.Add(item)
			Next item
			cbPhrase.SelectedItem = selected
			If cbPhrase.Items.Count > 0 AndAlso cbPhrase.SelectedItem Is Nothing Then
				cbPhrase.SelectedIndex = 0
			End If
		Finally
			cbPhrase.EndUpdate()
		End Try
	End Sub

	Private Sub EnableControls()
		Dim isIdle As Boolean = Not IsBusy()
		Dim fromFile As Boolean = rbFromFile.Checked
		rbMicrophone.Enabled = _biometricClient.VoiceCaptureDevice IsNot Nothing AndAlso _biometricClient.VoiceCaptureDevice.IsAvailable AndAlso (_biometricClient.LocalOperations And NBiometricOperations.CreateTemplate) <> 0
		gbPhrase.Enabled = isIdle
		btnOpenFile.Enabled = fromFile AndAlso isIdle
		btnStop.Enabled = (Not fromFile) AndAlso Not isIdle
		btnStart.Enabled = (Not fromFile) AndAlso isIdle
		gbSource.Enabled = isIdle
		chnCaptureAutomatically.Visible = Not fromFile
		chnCaptureAutomatically.Enabled = isIdle
		btnForce.Visible = Not fromFile
		btnForce.Enabled = Not fromFile AndAlso Not chnCaptureAutomatically.Checked AndAlso Not isIdle

		Dim boldFinish As Boolean = isIdle AndAlso _voice IsNot Nothing AndAlso _voice.Status = NBiometricStatus.Ok
		btnFinish.Font = New Font(btnFinish.Font, If(boldFinish, FontStyle.Bold, FontStyle.Regular))

		busyIndicator.Visible = Not isIdle
	End Sub

	Private Sub SetIsBusy(ByVal value As Boolean)
		If value Then
			_isIdle.Reset()
		Else
			_isIdle.Set()
		End If
	End Sub

	Private Function IsBusy() As Boolean
		Return Not _isIdle.WaitOne(0)
	End Function

	Private Sub CancelAndWait()
		If IsBusy() Then
			_biometricClient.Cancel()
			_isIdle.WaitOne()
		End If
	End Sub

#End Region

#Region "Private events"

	Private Sub OnVoiceCaptureDeviceChanged()
		Dim device = _biometricClient.VoiceCaptureDevice
		If device Is Nothing OrElse Not device.IsAvailable Then
			rbFromFile.Checked = True
			rbMicrophone.Text = "Microphone (Not connected)"
		Else
			rbMicrophone.Text = String.Format("Microphone ({0})", device.DisplayName)
		End If
		EnableControls()
	End Sub

	Private Sub OnBiometricClientPropertyChanged(ByVal sender As Object, ByVal e As PropertyChangedEventArgs)
		If e.PropertyName = "VoiceCaptureDevice" Then
			BeginInvoke(New MethodInvoker(AddressOf InvokeOnVoiceCaptureDeviceChanged))
		End If
	End Sub

	Private Sub InvokeOnVoiceCaptureDeviceChanged()
		If IsPageShown Then
			OnVoiceCaptureDeviceChanged()
		End If
	End Sub

	Private Sub BtnEditClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnEdit.Click
		Using form As New EditPhrasesForm()
			form.Phrases = _phrases
			form.ShowDialog(Me)
			SettingsManager.Phrases = _phrases
			ListAllPhrases()
		End Using
	End Sub

	Private Sub BtnOpenFileClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnOpenFile.Click
		If openFileDialog.ShowDialog() = DialogResult.OK Then
			Dim phrase As Phrase = TryCast(cbPhrase.SelectedItem, Phrase)
			_voice.SoundBuffer = Nothing
			_voice.FileName = openFileDialog.FileName
			_voice.PhraseId = If(phrase IsNot Nothing, phrase.Id, 0)
			_voice.CaptureOptions = NBiometricCaptureOptions.None
			Dim biometricTask = _biometricClient.CreateTask(NBiometricOperations.Segment Or NBiometricOperations.CreateTemplate, _newSubject)
			SetIsBusy(True)
			_biometricClient.BeginPerformTask(biometricTask, AddressOf OnCreateTemplateCompleted, Nothing)
			SetHint(Color.Orange, "Extracting record. Please wait ...")
			EnableControls()
		End If
	End Sub

	Private Sub OnCreateTemplateCompleted(ByVal r As IAsyncResult)
		Dim status As NBiometricStatus = NBiometricStatus.InternalError
		Try
			Dim biometricTask = _biometricClient.EndPerformTask(r)
			status = biometricTask.Status
			If biometricTask.Error IsNot Nothing Then
				Utilities.ShowError(biometricTask.Error)
			End If
		Catch ex As Exception
			Utilities.ShowError(ex)
		End Try

		SetIsBusy(False)
		BeginInvoke(New Action(Of NBiometricStatus)(AddressOf UpdateStatus), status)

	End Sub

	Private Sub UpdateStatus(ByVal status As NBiometricStatus)
		If IsPageShown Then
			If status = NBiometricStatus.Ok Then
				SetHint(Color.Green, "Extraction successful")
			Else
				SetHint(Color.Red, "Extraction failed: {0}", status)
			End If
			EnableControls()
		End If
	End Sub

	Private Sub BtnStartClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnStart.Click
		Dim phrase As Phrase = TryCast(cbPhrase.SelectedItem, Phrase)
		_voice.FileName = Nothing
		_voice.SoundBuffer = Nothing
		_voice.PhraseId = If(phrase IsNot Nothing, phrase.Id, 0)
		_voice.CaptureOptions = If(chnCaptureAutomatically.Checked, NBiometricCaptureOptions.None, NBiometricCaptureOptions.Manual)
		Dim biometricTask = _biometricClient.CreateTask(NBiometricOperations.Capture Or NBiometricOperations.Segment Or NBiometricOperations.CreateTemplate, _newSubject)
		SetIsBusy(True)
		_biometricClient.BeginPerformTask(biometricTask, AddressOf OnCreateTemplateCompleted, Nothing)
		SetHint(Color.Orange, "Extracting record. Please say phrase ...")
		EnableControls()
	End Sub

	Private Sub BtnStopClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnStop.Click
		btnStop.Enabled = False
		_biometricClient.Force()
	End Sub

	Private Sub RadioButtonCheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
		If (CType(sender, RadioButton)).Checked Then
			ToggleRadioButtons()
		End If
	End Sub

	Private Sub ToggleRadioButtons()
		EnableControls()
	End Sub

	Private Sub BtnFinishClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnFinish.Click
		PageController.NavigateToStartPage()
	End Sub

	Private Sub BtnForceClick(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnForce.Click
		_biometricClient.ForceStart()
		btnForce.Enabled = False
	End Sub

#End Region

End Class

Public Class Phrase
#Region "Public fields"

	Public ReadOnly Id As Integer
	Public ReadOnly [String] As String

#End Region

#Region "Public constructor"

	Public Sub New(ByVal id As Integer, ByVal phrase As String)
		Me.Id = id
		[String] = phrase
	End Sub

#End Region

#Region "public methods"

	Public Overrides Function ToString() As String
		Return [String]
	End Function

	Public Overloads Overrides Function Equals(ByVal obj As Object) As Boolean
		Dim ph As Phrase = TryCast(obj, Phrase)
		If ph IsNot Nothing Then
			Return String.Equals([String], ph.String) AndAlso Id = ph.Id
		End If
		Return MyBase.Equals(obj)
	End Function

	Public Overrides Function GetHashCode() As Integer
		Return MyBase.GetHashCode()
	End Function

#End Region
End Class

File: java/Samples/Biometrics/CS/AbisSampleCS/SubjectEditor/CaptureVoicePage.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Threading;
using System.Windows.Forms;
using Neurotec.Biometrics;
using Neurotec.Biometrics.Client;

namespace Neurotec.Samples
{
	public partial class CaptureVoicePage : Neurotec.Samples.PageBase
	{
		#region Public constructor

		public CaptureVoicePage()
		{
			InitializeComponent();
		}

		#endregion

		#region Private fields

		private NBiometricClient _biometricClient;
		private NSubject _subject;
		private NSubject _newSubject;
		private NVoice _voice;
		private ManualResetEvent _isIdle = new ManualResetEvent(true);
		private List<Phrase> _phrases = new List<Phrase>();

		#endregion

		#region Public methods

		public override void OnNavigatedTo(params object[] args)
		{
			if (args == null || args.Length != 2) throw new ArgumentException("args");

			_subject = (NSubject)args[0];
			_biometricClient = (NBiometricClient)args[1];
			_biometricClient.PropertyChanged += OnBiometricClientPropertyChanged;

			_newSubject = new NSubject();
			_voice = new NVoice();
			_newSubject.Voices.Add(_voice);
			voiceView.Voice = _voice;
			_phrases = new List<Phrase>(SettingsManager.Phrases);
			ListAllPhrases();
			lblHint.Visible = false;

			OnVoiceCaptureDeviceChanged();
			EnableControls();

			base.OnNavigatedTo(args);
		}

		public override void OnNavigatingFrom()
		{
			if (IsBusy()) LongActionDialog.ShowDialog(this, "Finishin current action ...", CancelAndWait);
			_biometricClient.PropertyChanged -= OnBiometricClientPropertyChanged;
			if (_voice.Status == NBiometricStatus.Ok)
			{
				var voices = _newSubject.Voices.ToArray();
				_newSubject.Clear();
				foreach (var item in voices)
				{
					_subject.Voices.Add(item);
				}
			}
			voiceView.Voice = null;
			_newSubject = null;
			_voice = null;
			_subject = null;
			_biometricClient = null;

			base.OnNavigatingFrom();
		}

		#endregion

		#region Private methods

		private void SetHint(Color backColor, string format, params object[] args)
		{
			lblHint.BackColor = backColor;
			lblHint.Text = string.Format(format, args);
			lblHint.Visible = true;
		}

		private void ListAllPhrases()
		{
			try
			{
				cbPhrase.BeginUpdate();
				object selected = cbPhrase.SelectedItem;
				cbPhrase.Items.Clear();
				foreach (var item in _phrases)
				{
					cbPhrase.Items.Add(item);
				}
				cbPhrase.SelectedItem = selected;
				if (cbPhrase.Items.Count > 0 && cbPhrase.SelectedItem == null)
				{
					cbPhrase.SelectedIndex = 0;
				}
			}
			finally
			{
				cbPhrase.EndUpdate();
			}
		}

		private void EnableControls()
		{
			bool isIdle = !IsBusy();
			bool fromFile = rbFromFile.Checked;
			rbMicrophone.Enabled = _biometricClient.VoiceCaptureDevice != null &&
				_biometricClient.VoiceCaptureDevice.IsAvailable &&
				(_biometricClient.LocalOperations & (NBiometricOperations.CreateTemplate | NBiometricOperations.Segment)) != 0;
			gbPhrase.Enabled = isIdle;
			btnOpenFile.Enabled = fromFile && isIdle;
			btnStop.Enabled = !fromFile && !isIdle;
			btnStart.Enabled = !fromFile && isIdle;
			gbSource.Enabled = isIdle;
			chbCaptureAutomatically.Visible = !fromFile;
			chbCaptureAutomatically.Enabled = isIdle;
			btnForce.Visible = !fromFile;
			btnForce.Enabled = !fromFile && !chbCaptureAutomatically.Checked && !isIdle;

			bool boldFinish = isIdle && _voice != null && _voice.Status == NBiometricStatus.Ok;
			btnFinish.Font = new Font(btnFinish.Font, boldFinish ? FontStyle.Bold : FontStyle.Regular);

			busyIndicator.Visible = !isIdle;
		}

		private void SetIsBusy(bool value)
		{
			if (value)
				_isIdle.Reset();
			else
				_isIdle.Set();
		}

		private bool IsBusy()
		{
			return !_isIdle.WaitOne(0);
		}

		private void CancelAndWait()
		{
			if (IsBusy())
			{
				_biometricClient.Cancel();
				_isIdle.WaitOne();
			}
		}

		private void OnVoiceCaptureDeviceChanged()
		{
			var device = _biometricClient.VoiceCaptureDevice;
			if (device == null || !device.IsAvailable)
			{
				rbFromFile.Checked = true;
				rbMicrophone.Text = "Microphone (Not connected)";
			}
			else
			{
				rbMicrophone.Text = string.Format("Microphone ({0})", device.DisplayName);
			}
			EnableControls();
		}

		private void OnBiometricClientPropertyChanged(object sender, System.ComponentModel.PropertyChangedEventArgs e)
		{
			if (e.PropertyName == "VoiceCaptureDevice")
			{
				BeginInvoke(new Action(() =>
				{
					if (IsPageShown) OnVoiceCaptureDeviceChanged();
				}));
			}
		}

		#endregion

		#region Private events

		private void BtnEditClick(object sender, EventArgs e)
		{
			using (EditPhrasesForm form = new EditPhrasesForm())
			{
				form.Phrases = _phrases;
				form.ShowDialog(this);
				SettingsManager.Phrases = _phrases;
				ListAllPhrases();
			}
		}

		private void BtnOpenFileClick(object sender, EventArgs e)
		{
			if (openFileDialog.ShowDialog() == DialogResult.OK)
			{
				Phrase phrase = cbPhrase.SelectedItem as Phrase;
				_voice.SoundBuffer = null;
				_voice.FileName = openFileDialog.FileName;
				_voice.PhraseId = phrase != null ? phrase.Id : 0;
				var biometricTask = _biometricClient.CreateTask(NBiometricOperations.Segment | NBiometricOperations.CreateTemplate, _newSubject);
				SetIsBusy(true);
				_biometricClient.BeginPerformTask(biometricTask, OnCreateTemplateCompleted, null);
				SetHint(Color.Orange, "Extracting record. Please wait ...");
				EnableControls();
			}
		}

		private void OnCreateTemplateCompleted(IAsyncResult r)
		{
			NBiometricStatus status = NBiometricStatus.InternalError;
			try
			{
				var biometricTask = _biometricClient.EndPerformTask(r);
				status = biometricTask.Status;
				if (biometricTask.Error != null) Utilities.ShowError(biometricTask.Error);
			}
			catch (Exception ex)
			{
				Utilities.ShowError(ex);
			}

			SetIsBusy(false);
			BeginInvoke(new Action<NBiometricStatus>(taskStatus =>
			{
				if (IsPageShown)
				{
					if (taskStatus == NBiometricStatus.Ok)
						SetHint(Color.Green, "Extraction successful");
					else
						SetHint(Color.Red, "Extraction failed: {0}", taskStatus);
					EnableControls();
				}
			}), status);
		}

		private void BtnStartClick(object sender, EventArgs e)
		{
			Phrase phrase = cbPhrase.SelectedItem as Phrase;
			_voice.FileName = null;
			_voice.SoundBuffer = null;
			_voice.PhraseId = phrase != null ? phrase.Id : 0;
			_voice.CaptureOptions = chbCaptureAutomatically.Checked ? NBiometricCaptureOptions.None : NBiometricCaptureOptions.Manual;
			var biometricTask = _biometricClient.CreateTask(NBiometricOperations.Capture | NBiometricOperations.Segment | NBiometricOperations.CreateTemplate, _newSubject);
			SetIsBusy(true);
			_biometricClient.BeginPerformTask(biometricTask, OnCreateTemplateCompleted, null);
			SetHint(Color.Orange, "Extracting record. Please say phrase ...");
			EnableControls();
		}

		private void BtnStopClick(object sender, EventArgs e)
		{
			btnStop.Enabled = false;
			_biometricClient.Force();
		}

		private void RadioButtonCheckedChanged(object sender, EventArgs e)
		{
			if (((RadioButton)sender).Checked)
				ToggleRadioButtons();
		}

		private void ToggleRadioButtons()
		{
			EnableControls();
		}

		private void BtnFinishClick(object sender, EventArgs e)
		{
			PageController.NavigateToStartPage();
		}

		private void BtnForceClick(object sender, EventArgs e)
		{
			_biometricClient.ForceStart();
			btnForce.Enabled = false;
		}

		#endregion
	}

	public class Phrase
	{
		#region Public fields

		public readonly int Id;
		public readonly string String;

		#endregion

		#region Public constructor

		public Phrase(int id, string phrase)
		{
			Id = id;
			String = phrase;
		}

		#endregion

		#region public methods

		public override string ToString()
		{
			return String;
		}

		public override bool Equals(object obj)
		{
			Phrase ph = obj as Phrase;
			if (ph != null)
			{
				return string.Equals(String, ph.String) && Id == ph.Id;
			}
			return base.Equals(obj);
		}

		public override int GetHashCode()
		{
			return base.GetHashCode();
		}

		#endregion
	}
}
